### PR TITLE
feat: barrnap-wrapper

### PR DESCRIPTION
**Pull request by currocam**
### Description
A wrapper for barrnap, a  bacterial ribosomal RNA predictor.

### QC

* [ ] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

### CODE REVIEW

File: bio/barrnap/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+valid_keys = ["gff", "fasta"]
+for key in snakemake.output.keys():
+    assert (
+        key in valid_keys
+    ), "Invalid key in output. Valid keys are: %r. Given: %r." % (valid_keys, key)
+
+gff_out = snakemake.output.get("gff")
+assert gff_out, "Output must contain at least a gff file."
+
```
Comment (fgvieira) [inline bio/barrnap/wrapper.py:31]: This part is not really necessary.

```suggestion
```

File: bio/barrnap/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+if fasta_out:
+    extra += f" -o {fasta_out}"
+
+shell(
+    "barrnap"
+    " --threads {snakemake.threads}"
+    " -k {kingdom}"
+    " {extra}"
+    " < {snakemake.input}"
+    " > {gff_out}"
```
Comment (fgvieira) [inline bio/barrnap/wrapper.py:42]:
```suggestion
    " > {snakemake.output.gff}"
```

File: bio/barrnap/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+fasta_out = snakemake.output.get("fasta")
+if fasta_out:
+    extra += f" -o {fasta_out}"
+
+shell(
+    "barrnap"
+    " --threads {snakemake.threads}"
+    " -k {kingdom}"
+    " {extra}"
+    " < {snakemake.input}"
```
Comment (fgvieira) [inline bio/barrnap/wrapper.py:41]:
```suggestion
    " < {snakemake.input.fasta}"
```

File: bio/barrnap/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+assert gff_out, "Output must contain at least a gff file."
+
+fasta_out = snakemake.output.get("fasta")
+if fasta_out:
+    extra += f" -o {fasta_out}"
+
+shell(
+    "barrnap"
+    " --threads {snakemake.threads}"
+    " -k {kingdom}"
```
Comment (fgvieira) [inline bio/barrnap/wrapper.py:39]:
```suggestion
    " -k {snakemake.params.kingdom}"

```

File: bio/barrnap/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+extra = snakemake.params.get("extra", "")
+kingdom = snakemake.params.get("kingdom", "")
+assert kingdom in [
+    "bac",
+    "arc",
+    "euk",
+    "mito",
+], "Invalid kingdom. Valid kingdoms are: 'bac', 'arc', 'euk', 'mito'. Given: %r." % (
+    kingdom
+)
```
Comment (fgvieira) [inline bio/barrnap/wrapper.py:21]: The program will fail if the kingdom is not one of those, right?
If so, this part is not necessary.

Comment (currocam) [reply bio/barrnap/wrapper.py:21]: Yes, you're right

Comment (currocam) [reply bio/barrnap/wrapper.py]: I just added the default kingdom (the one you use if there's no --kingdom flag)

File: bio/barrnap/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+__copyright__ = "Copyright 2023, Curro Campuzano Jiménez"
+__email__ = "campuzanocurro@gmail.com"
+__license__ = "MIT"
+
+
+from snakemake.shell import shell
+
+log = snakemake.log_fmt_shell(stdout=False, stderr=True)
+extra = snakemake.params.get("extra", "")
+kingdom = snakemake.params.get("kingdom", "bac")
```
Comment (fgvieira) [inline bio/barrnap/wrapper.py:13]:
```suggestion
```

Comment (currocam) [reply bio/barrnap/wrapper.py:13]: Actually, it was a typo. I was trying to use it in the shell command with --kingdom. With the default "bac", if no kingdom parameter is given in the params, the program runs as if no --kingdom paramater was given (so as expected)

**Review by fgvieira — APPROVED** [review]

**Pull request merged**